src/algebra/free_algebra.lean

```diff
@@ -0,0 +1,293 @@
+  map_zero' := rfl,
+  map_add' := λ _ _, quot.sound rel.add_scalar,
+  commutes' := λ _, by { rintros ⟨⟩, exact quot.sound rel.central_scalar },
+  smul_def' := λ _ _, rfl }
+
+/--
+The canonical function `X → free_algebra R X`.
+-/
+def ι : X → free_algebra R X :=
+λ m, quot.mk _ m
+
+@[simp] lemma quot_mk_eq_ι (m : X) : quot.mk (free_algebra.rel R X) m = ι R X m := rfl
+
+/--
+Given a function `f : X → A` where `A` is an `R`-algebra, `lift R X f` is the unique lift
+of `f` to a morphism of `R`-algebras `free_algebra R X → A`.
+-/
+def lift {A : Type*} [semiring A] [algebra R A] (f : X → A) : free_algebra R X →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { exact (algebra_map R A).map_add h_r h_s, },
+    { exact (algebra_map R A).map_mul h_r h_s },
+    { apply algebra.commutes },
+    { change _ + _ + _ = _ + (_ + _),
+      rw add_assoc },
+    { change _ + _ = _ + _,
+      rw add_comm, },
+    { change (algebra_map _ _ _) + lift_fun R X f _ = lift_fun R X f _,
+      simp, },
+    { change _ * _ * _ = _ * (_ * _),
+      rw mul_assoc },
+    { change (algebra_map _ _ _) * lift_fun R X f _ = lift_fun R X f _,
+      simp, },
+    { change lift_fun R X f _ * (algebra_map _ _ _) = lift_fun R X f _,
+      simp, },
+    { change _ * (_ + _) = _ * _ + _ * _,
+      rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _,
+      rw right_distrib, },
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _,
+      simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _,
+      simp },
+    repeat { change lift_fun R X f _ + lift_fun R X f _ = _,
+      rw h_ih,
+      refl, },
+    repeat { change lift_fun R X f _ * lift_fun R X f _ = _,
+      rw h_ih,
+      refl, },
+  end,
+  map_one' := by { change algebra_map _ _ _ = _, simp },
+  map_mul' := by { rintros ⟨⟩ ⟨⟩, refl },
+  map_zero' := by { change algebra_map _ _ _ = _, simp },
+  map_add' := by { rintros ⟨⟩ ⟨⟩, refl },
+  commutes' := by tauto }
+
+variables {R X}
+
+@[simp]
+theorem ι_comp_lift {A : Type*} [semiring A] [algebra R A] (f : X → A) :
+  (lift R X f : free_algebra R X → A) ∘ (ι R X) = f := by {ext, refl}
+
+@[simp]
+theorem lift_ι_apply {A : Type*} [semiring A] [algebra R A] (f : X → A) (x) :
+  lift R X f (ι R X x) = f x := rfl
+
+@[simp]
+theorem lift_unique {A : Type*} [semiring A] [algebra R A] (f : X → A)
+  (g : free_algebra R X →ₐ[R] A) :
+  (g : free_algebra R X → A) ∘ (ι R X) = f ↔ g = lift R X f :=
+begin
+  refine ⟨λ hyp, _, λ hyp, by rw [hyp, ι_comp_lift]⟩,
+  ext,
+  rcases x,
+  induction x,
+  { change ((g : free_algebra R X → A) ∘ (ι R X)) _ = _,
+    rw hyp,
+    refl },
+  { exact alg_hom.commutes g x },
+  { change g (quot.mk _ _ + quot.mk _ _) = _,
+    rw [alg_hom.map_add, x_ih_a, x_ih_a_1],
+    refl },
+  { change g (quot.mk _ _ * quot.mk _ _) = _,
+    rw [alg_hom.map_mul, x_ih_a, x_ih_a_1],
+    refl },
+end
+
+attribute [irreducible] free_algebra ι lift
```

Comment:
I completely agree with this.
Should we add a comment explaining that people should rely entirely on the universal property?

Reply:
```suggestion
/-!
At this stage we set the basic definitions as `@[irreducible]`, so from this point onwards one should only use the universal properties of the free algebra, and consider the actual implementation as a quotient of an inductive type as completely hidden.

Of course, one still has the option to locally make these definitions `semireducible` if so desired, and Lean is still willing in some circumstances to do unification based on the underlying definition.
-/
attribute [irreducible] free_algebra ι lift
```

Reply:
Done!